packages/installer/src/dappGet/fetch/DappGetFetcher.ts

```diff
@@ -78,7 +71,85 @@ export class DappGetFetcher {
           .filter((version) => satisfies(version, versionRange));
       }
     }
-    // Case 3. unvalid semver version ("/ipfs/Qmre4..."): Asume it's the only version
+    // Case 3. invalid semver version ("/ipfs/Qmre4..."): Assume it's the only version
     return [versionRange];
   }
-}
+
+  /**
+   * Merges optional dependencies into the main dependencies object if the corresponding 
+   * packages are installed.
+   *
+   * @param dependencies The main dependencies object to be merged into.
+   * @param optionalDependencies The optional dependencies to be checked and merged.
+   * @param installedPackages The list of currently installed packages.
+   */
+  private mergeOptionalDependencies(
+    dependencies: Dependencies,
+    optionalDependencies: Dependencies,
+    installedPackages: InstalledPackageData[]
+  ): void {
+    for (const [optionalDepName, optionalDepVersion] of Object.entries(optionalDependencies)) {
+      const isInstalled = installedPackages.some(
+        (installedPackage) => installedPackage.dnpName === optionalDepName
+      );
+
+      if (isInstalled) {
+        dependencies[optionalDepName] = optionalDepVersion;
+      }
+    }
+  }
+
+  /**
+   * Processes the dependencies by first filtering out those that are already satisfied by installed packages
+   * and then converting any remaining semver ranges to appropriate APM versions.
+   *
+   * @param dependencies The main dependencies object to be processed.
+   * @param installedPackages The list of currently installed packages.
+   */
+  private filterSatisfiedDependencies(
+    dependencies: Dependencies,
+    installedPackages: InstalledPackageData[]
+  ): void {
+    for (const [depName, depVersion] of Object.entries(dependencies)) {
+      const installedPackage = installedPackages.find(
+        (pkg) => pkg.dnpName === depName
+      );
+
+      if (!validRange(depVersion))
+        throw new Error(`Invalid semver notation for dependency ${depName}: ${depVersion}`);
+
+      // Remove dependency if it is already satisfied by an installed package
+      if (installedPackage && satisfies(installedPackage.version, depVersion)) {
+        console.log(
+          `Dependency ${depName} is already installed with version ${installedPackage.version}`
+        );
+        delete dependencies[depName];
+      }
+    }
+  }
+
+  private async defineExactVersions(
```

Comment:
could you add a description to the function also providing an example of how looks the object before and after?

Reply:
Yes! Done